run-make-support: adjust assertion printing, add some basic sanity checks

jieyouxu opened the pull request:
cc @Noratrieb

I think we may have unintentionally regressed this recently and double-printed (or printed even when the assertions didn't fail). This PR should condition the detail dumps only when the assertions fail.

Added some basic sanity checks for the assertions helpers except for the directory comparisons. That particular helper is not robust against symlinks, and I intend to address it in a follow-up (issue is #135037).

r? bootstrap (or compiler)

compiler-errors commented on src/tools/run-make-support/src/assertion_helpers/mod.rs:
```diff
@@ -0,0 +1,160 @@
+        eprintln!("{}:", self.assertion_name);
+        eprintln!("=== HAYSTACK ===");
+        eprintln!("{}", self.haystack);
+        eprintln!("=== {} ===", self.query_kind);
```
This is never not "NEEDLE". Why abstract it?

jieyouxu replied:
See the regex variants (e.g. `assert_contains_regex` and friends), this can actually be `query_kind: "REGEX",`

compiler-errors commented on src/tools/run-make-support/src/assertion_helpers/mod.rs:
```diff
@@ -0,0 +1,160 @@
+pub fn assert_count_is<H: AsRef<str>, N: AsRef<str>>(
+    expected_count: usize,
+    haystack: H,
+    pattern: N,
```
Renaming `pattern` to `needle` but only for some of these functions is a bit inconsistent 🤔

jieyouxu replied:
This was intended to distinguish between the assert helpers that accept a plain-text substring vs regex patterns.

jieyouxu replied:
Reverted these renames, just used `needle` consistently.

compiler-errors reviewed (approved):
r=me with or without nits

compiler-errors commented:
this was very hard to diff btw lol, i had to open it in a separate diff tool 💀

jieyouxu commented:
Yeah sorry, I cooked this up hastly, should've made this into several commits.

jieyouxu commented:
If you prefer, I can properly split the commits tmrw. Actually, I'm going to do that if only to make the git blame on the assertion helpers not ass.

jieyouxu committed:
run-make-support: convert `assertion_helpers` to module

jieyouxu committed:
run-make-support: add basic sanity tests for assertion helpers

jieyouxu committed:
run-make-support: tidy up assertion failure dumping

Avoid double-dumping or dumping even when assertion is successful.

jieyouxu commented:
Changes since last review:

- Actually properly split the changes into 3 commits:
    1. First commit only moves `assertion_helpers.rs` to `assertion_helpers/mod.rs`.
    2. Second commit introduces the basic sanity tests.
    3. Third commit fixes the assertion dump double-printing, avoids dumping even when the assertion is successful, and slightly tidies up the dump handling.
 - Added a FIXME backlinking to #135037 for the dir assertion helper.

There is no functional changes since last review, I just tidied up the commit history to make the diff less of a pain to look at or blame.

@rustbot review

jieyouxu commented:
Split into several commits and addressed the nits, so

@bors r=compiler-errors rollup

bors commented:
:pushpin: Commit 6175d738625d07a1e3d3c7db06b6d6a69f59a04c has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 6175d738625d07a1e3d3c7db06b6d6a69f59a04c -->
<!-- homu: {"type":"Approved","sha":"6175d738625d07a1e3d3c7db06b6d6a69f59a04c","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
